compiler/noirc_evaluator/src/ssa/opt/basic_conditional.rs

```diff
@@ -0,0 +1,526 @@
+                    block_entry: block,
+                    block_then: Some(*then_destination),
+                    block_else: Some(*else_destination),
+                    block_exit: next_left.unwrap(),
+                });
+            }
+        }
+    } else if left_successors_len == 1 && next_left == Some(right) {
+        // Left branch joins the right branch, e.g if/then statement with no else
+        // I am not sure whether this case can happen, but it is not difficult to handle it
+        let cost = block_cost(left, &function.dfg);
+        if cost < cost / 2 + jump_overhead {
+            if let Some(TerminatorInstruction::JmpIf {
+                condition: _,
+                then_destination,
+                else_destination,
+                call_stack: _,
+            }) = function.dfg[block].terminator()
+            {
+                let (block_then, block_else) = if left == *then_destination {
+                    (Some(left), None)
+                } else if left == *else_destination {
+                    (None, Some(left))
+                } else {
+                    return None;
+                };
+
+                result = Some(BasicConditional {
+                    block_entry: block,
+                    block_then,
+                    block_else,
+                    block_exit: right,
+                });
+            }
+        }
+    } else if right_successors_len == 1 && next_right == Some(left) {
+        // Right branch joins the left branch, e.g if/else statement with no then
+        // I am not sure whether this case can happen, but it is not difficult to handle it
+        let cost = block_cost(right, &function.dfg);
+        if cost < cost / 2 + jump_overhead {
+            if let Some(TerminatorInstruction::JmpIf {
+                condition: _,
+                then_destination,
+                else_destination,
+                call_stack: _,
+            }) = function.dfg[block].terminator()
+            {
+                let (block_then, block_else) = if right == *then_destination {
+                    (Some(right), None)
+                } else if right == *else_destination {
+                    (None, Some(right))
+                } else {
+                    return None;
+                };
+                result = Some(BasicConditional {
+                    block_entry: block,
+                    block_then,
+                    block_else,
+                    block_exit: right,
+                });
+            }
+        }
+    }
+    // A conditional exit would have exactly 2 predecessors
+    result.filter(|result| cfg.predecessors(result.block_exit).len() == 2)
+}
+
+/// Computes a cost estimate of a basic block
+/// returns u32::MAX if the block has side-effect instructions
+/// WARNING: the estimates are estimate of the runtime cost of each instructions,
+/// 1 being the cost of the simplest instruction. These numbers can be improved.
```

Comment:
Let's make an issue for tracking purposes

Reply:
Done here: #7543